custom/ewsghana/reports/specific_reports/stock_status_report.py

```diff
@@ -382,6 +339,156 @@ class StockStatus(MultiReport):
     is_exportable = True
 
     @property
+    def get_supply_points(self):
+        supply_points = []
+        if self.report_config['location_id']:
+            location = SQLLocation.objects.get(
+                domain=self.report_config['domain'],
+                location_id=self.report_config['location_id']
+            )
+            if location.location_type.name == 'country':
+                supply_points = SQLLocation.objects.filter(
+                    Q(parent__location_id=self.report_config['location_id'], is_archived=False) |
+                    Q(location_type__name='Regional Medical Store', domain=self.report_config['domain']) |
+                    Q(location_type__name='Teaching Hospital', domain=self.report_config['domain'])
+                ).order_by('name').exclude(supply_point_id__isnull=True)
+            else:
+                supply_points = SQLLocation.objects.filter(
+                    parent__location_id=self.report_config['location_id'], is_archived=False,
+                    location_type__administrative=False,
+                ).order_by('name').exclude(supply_point_id__isnull=True)
+        return supply_points
+
+    def unique_products(self, locations, all=False):
+        if self.report_config['products'] and not all:
+            return SQLProduct.objects.filter(
+                pk__in=locations.values_list('_products', flat=True),
+            ).filter(pk__in=self.report_config['products']).exclude(is_archived=True)
+        elif self.report_config['program'] and not all:
+            return SQLProduct.objects.filter(
+                pk__in=locations.values_list('_products', flat=True),
+                program_id=self.report_config['program']
+            ).exclude(is_archived=True)
+        else:
+            return SQLProduct.objects.filter(
+                pk__in=locations.values_list('_products', flat=True)
+            ).exclude(is_archived=True)
```

Comment:
This method looks like it could be a big performance bottleneck.  `_products` is a many-to-many field, what does the `flat` return value of that look like?  There are several querysets being evaluated and passed to other querysets.  Using relational lookups would likely be more straightforward and performant.  Can you separate this out to a dbaccessor-type function with clearly-defined inputs and outputs?

Also, this could probably be simplified by using the reverse relationship of the `_products` many-to-many field.